src/Grpc.Net.Client/Internal/Retry/RetryCallBase.cs

```diff
@@ -0,0 +1,504 @@
+
+            lock (Lock)
+            {
+                if (!BufferedCurrentMessage)
+                {
+                    messageData = SerializePayload(call, callOptions, message);
+
+                    // Don't buffer message data if the call has been commited.
+                    if (!CommitedCallTask.IsCompletedSuccessfully())
+                    {
+                        if (!TryAddToRetryBuffer(messageData))
+                        {
+                            CommitCall(call, CommitReason.BufferExceeded);
+                        }
+                        else
+                        {
+                            BufferedCurrentMessage = true;
+
+                            Log.MessageAddedToBuffer(Logger, messageData.Length, CurrentCallBufferSize);
+                        }
+                    }
+                }
+                else
+                {
+                    // There is a race between:
+                    // 1. A client stream starting for a new call. It will write all buffered messages, and
+                    // 2. Writing a new message here. The message may already have been buffered when the client
+                    //    stream started so we don't want to write it again.
+                    //
+                    // Check the client stream write count against he buffer message count to ensure all buffered
+                    // messages haven't already been written.
+                    if (call.MessagesWritten == BufferedMessages.Count)
+                    {
+                        return;
+                    }
+
+                    messageData = BufferedMessages[BufferedMessages.Count - 1];
+                }
+            }
+
+            await call.WriteMessageAsync(writeStream, messageData, callOptions.CancellationToken).ConfigureAwait(false);
+        }
+
+        protected void CommitCall(IGrpcCall<TRequest, TResponse> call, CommitReason commitReason)
+        {
+            lock (Lock)
+            {
+                if (!CommitedCallTask.IsCompletedSuccessfully())
+                {
+                    // The buffer size is verified in unit tests after calls are completed.
+                    // Clear the buffer before commiting call.
+                    ClearRetryBuffer();
+
+                    OnCommitCall(call);
+
+                    // Log before committing for unit tests.
+                    Log.CallCommited(Logger, commitReason);
+
+                    NewActiveCallTcs?.SetResult(null);
+                    _commitedCallTcs.SetResult(call);
+                }
+            }
+        }
+
+        protected abstract void OnCommitCall(IGrpcCall<TRequest, TResponse> call);
+
+        protected bool HasClientStream()
+        {
+            return Method.Type == MethodType.ClientStreaming || Method.Type == MethodType.DuplexStreaming;
+        }
+
+        protected void SetNewActiveCallUnsynchronized(IGrpcCall<TRequest, TResponse> call)
```

Comment:
Might want to mention in a comment that this call can only be made when holding a lock.

Reply:
The Unsynchronized suffix does that.